src/providers/exoscale/mod.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2020 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! metadata fetcher for the exoscale provider
+
+use std::collections::HashMap;
+
+use openssh_keys::PublicKey;
+use slog_scope::warn;
+
+use crate::errors::*;
+use crate::network;
+use crate::providers::MetadataProvider;
+use crate::retry;
+
+#[derive(Clone, Debug)]
+pub struct ExoscaleProvider {
+    client: retry::Client,
+}
+
+impl ExoscaleProvider {
+    pub fn try_new() -> Result<ExoscaleProvider> {
+        let client = retry::Client::try_new()?;
+
+        Ok(ExoscaleProvider {
+            client,
+        })
+    }
+
+    fn endpoint_for(&self, key: &str) -> String {
+        format!("http://169.254.169.254/1.0/meta-data/{}", key)
+    }
+}
+
+impl MetadataProvider for ExoscaleProvider {
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        let mut out = HashMap::with_capacity(10);
+        let add_value = |map: &mut HashMap<_, _>, key: &str, name| -> Result<()> {
+            let value = self
+                .client
+                .get(retry::Raw, self.endpoint_for(name))
+                .send()?;
+
+            if let Some(value) = value {
+                map.insert(key.to_string(), value);
+            }
+
+            Ok(())
+        };
+
+        add_value(&mut out, "EXOSCALE_INSTANCE_ID", "instance-id")?;
+        add_value(&mut out, "EXOSCALE_LOCAL_HOSTNAME", "local-hostname")?;
+        add_value(&mut out, "EXOSCALE_PUBLIC_HOSTNAME", "public-hostname")?;
+        add_value(
+            &mut out,
+            "EXOSCALE_AVAILABILITY_ZONE",
+            "availability-zone",
+        )?;
+        add_value(&mut out, "EXOSCALE_IPV4_PUBLIC", "public-ipv4")?;
+        add_value(&mut out, "EXOSCALE_IPV4_LOCAL", "local-ipv4")?;
+        add_value(&mut out, "EXOSCALE_SERVICE_OFFERING", "service-offering")?;
+        add_value(&mut out, "EXOSCALE_CLOUD_IDENTIFIER", "cloud-identifier")?;
+        add_value(&mut out, "EXOSCALE_VM_ID", "vm-id")?;
+        add_value(&mut out, "EXOSCALE_SERVICE_OFFERING", "service-offering")?;
+
+        Ok(out)
+    }
+
+    fn hostname(&self) -> Result<Option<String>> {
+        Ok(None)
```

Comment:
It looks like the hostname is actually available through the metadata.

However I'm not sure what's the difference between "local hostname" and "public hostname", and which one would fit better here.

Reply:
Thanks! I had in mind to add it but have to watch which one to pick-up (public/private)

Reply:
I'll fix that today 😉

Reply:
I added the implementation for the hostname thanks

Reply:
is it ok for you?

Reply:
:+1: